src/wmtk/PolyMesh.hpp

```diff
@@ -0,0 +1,264 @@
+ *
+ * Also note that the natural generalization of this implementation is the facet-edge data
+ * structure of Dobkin and Laszlo, where the primitive is an oriented face-edge
+ * triple. However, while the halfedge can be encoded as an edge attribute, the facet-edge
+ * data needs to be encoded on actual pairs of faces and edges as each face can be incident
+ * to many edges and vice versa.
+ */
+
+/**
+ * Helper Functions
+ *
+ * TODO: These can be put in another file, e.g., polymesh_topology_initialization, to conform
+ * with the code organization used for trimesh and tetmesh. Or they can be made static member
+ * functions. Or they could be defined only in the cpp file.
+ */
+
+/**
+ * Convert from matrix (F) mesh representation to NOB data structure
+ *
+ * @param F: #f*n, each row represents the vertex id (ccw) of current face
+ * @return next_he: (N) size #h vector, next halfedge id
+ * @return opp: (O) size #h vector, opposite halfedge id
+ * @return bnd_loops: (B) collection of boundary face ids.
+ */
+std::tuple<VectorXl, VectorXl, VectorXl> fv_to_nob(Eigen::Ref<const RowVectors3l> F);
+
+/**
+ * Extend next_he and opp to add extra halfedges along the boundaries.
+ *
+ * @param next_he: next-halfedge map same length as opp
+ * @param opp: halfedge map, for boundary halfedges -1
+ * @return next_he_ext: next_halfedge map, same length as opp_ext; newly added halfedges are linked
+ *                      into boundary loops
+ * @return opp_ext: opp with an extra coupled halfedge added at the end for each boundary halfedge
+ *                  all halfedges have a pair
+ */
+std::tuple<VectorXl, VectorXl> build_boundary_loops(
+    Eigen::Ref<const VectorXl> next_he,
+    Eigen::Ref<const VectorXl> opp);
+
+/**
+ * Build orbits following next id recorded in perm.
+ *
+ * @param perm: a permutation given by a list of non-repeating integers in the range 0..len(perm)
+ * @return cycles: a list of lists, each list represents a cycle of perm
+ */
+std::vector<std::vector<int>> build_orbits(const std::vector<int>& perm);
+
+/**
+ * End of Helper Functions
+ */
+
+namespace wmtk {
+
+
+class PolyMesh : public Mesh
+{
+public:
+    PolyMesh();
+    PolyMesh(const PolyMesh& o);
+    PolyMesh(PolyMesh&& o);
+    PolyMesh& operator=(const PolyMesh& o);
+    PolyMesh& operator=(PolyMesh&& o);
+
+    // TODO This only makes sense for simplices as currently named
+    // It is also potentially ambiguous as our top type is a face, but our basic mesh primitive
+    // is a halfedge
+    PrimitiveType top_simplex_type() const override { return PrimitiveType::Face; }
+
+    /**
+     * @brief Primitive for changing connectivity without changing the number of edges.
+     * 
+     * Requires t and s to have consistent orientation
+     */
+    Tuple splice(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Primitive for adding a bubble component with a single edge.
+     */
+    Tuple make_edge();
```

Comment:
i think you're mimicing the split_edge / collapse_edge operation stuff here - and to be honest I do not think that it's a great idea.

The original intent from way back when was that the class implements the operation and returns a reference `Tuple` to help perform navigation for other tasks. The issue that arose is that each operation's return Tuple had a different semantic and so every consumer of these return Tuple had to keep track of which operation the return data came from and how to do navigation with respect to that tuple.

IMO these operations will/should provide per-operation classes to help identify information about what happened (which tuple was deleted? what simplices got modified? etc)

Furthermore these operations really shouldn't be called by anyone other than some Operation object - and once they only really get invoked by an Operation object then the implementation might as well be shoved to those objects instead of being left within the mesh struct (in theory these functions are all protected if you keep htem in the Mesh).

Reply:
So should all mesh operations (splice, make_edge, split_face, joint_vertex, etc.) be implemented instead as Operations akin to those in `operations/tri_mesh`?

Reply:
Being more akin to the `opeartoins/tri_mesh` is my personal preference as it leaves the PolygonMesh structure itself fairly simple, but we can discuss this with others. Right now in TriMesh/TetMesh we have these TriMeshOperationExecutor objects which really are the Operation implementations but in a weird place? If you prefer constructing the execution structures and returning the data from those objects that's fine as well - but having all of this data in an Operation object would have been easier imo. These Executor classes were in part developed because the Operation infrastructure wasn't there yet and should potentially be refactored.